source/agora/consensus/ValidatorSet.d

```diff
@@ -91,21 +95,40 @@ public class ValidatorSet
         if (auto reason = isInvalidReason(enroll, finder))
             return reason;
 
-        // check if already exists
-        if (this.hasEnrollment(enroll.utxo_key))
-            return "This validator is already enrolled";
-
         try
         {
-            () @trusted {
-                const ZeroDistance = 0;  // initial distance
-                this.db.execute("INSERT INTO validator_set " ~
-                    "(key, cycle_length, enrolled_height, " ~
-                    "distance, preimage) VALUES (?, ?, ?, ?, ?)",
-                    enroll.utxo_key.toString(), enroll.cycle_length,
-                    block_height.value, ZeroDistance,
-                    enroll.random_seed.toString());
-            }();
+            const ZeroDistance = 0;  // initial distance
+            const bool exist = this.hasEnrollment(enroll.utxo_key);
+            if (exist)
+            {
+                // this checks that the re-enrollment is requested past the half
+                // of the validator cycle.
+                if (block_height.value - this.getEnrolledHeight(enroll.utxo_key) <
+                    this.ReEnrollAvailableRound)
+                {
+                    log.info("Rejected re-enrollment is too early, " ~
+                        "Data was: {}", enroll);
+                    return "The re-enrollment is too early";
+                }
+
+                () @trusted {
+                    this.db.execute("UPDATE validator_set SET " ~
+                        "enrolled_height = ?, distance = ?, preimage = ? WHERE key = ?",
+                        block_height.value, ZeroDistance, enroll.random_seed.toString(),
+                        enroll.utxo_key.toString());
+                }();
```

Comment:
This is overriding an Enrollment before the Enrollment is over. That's something we should absolutely never do, and that is not what the issue is about. The issue is about the ability to re-enroll on the last block where a node is a validator, to avoid a 2-blocks downtime in validation. It does not **override** being a validator, it extends it.